### PR TITLE
Create our own logits processor for Xgrammar, add support for mlx

### LINKED ISSUE #1698: Add support for the MLXLM models for the Xgrammar backend
The `Xgrammar` backend currently relies on a logits precessor directly provided by the library. An issue is that this logits processor only works with `torch` tensors (what's used by the `Transformers` model). As a result, it cannot be used with the `MLXLM` models that use `mlx` tensors while `xgrammar` does provide support for `mlx`.

To fix that issue, we should use the lower-level API of the `xgrammar` library and create a logits processor for that backend ourselves. That way, we would be able to handle both `torch` and `mlx` tensors. This is already what we do for the `LLGuidance` backend, so it should be possible to draw inspiration from it.

Additionally, creating our own logits processor is needed to add extra features such as a thinking mode.

### PR DESCRIPTION
**Pull request by RobinPicard**
Closes #1698 

This is a dependency for adding support for thinking models (#1627) as we will need to add some custom logic to the logits prorcessor, so we must create our own.

The PR adds supports for `mlx` but not for `numpy` (so the `LlamaCpp` model is not supported) as it does not seem that Xgrammar currently supports `numpy` tensors. If that changes in the future or if we find a way to make it work, it should be easy to add as the logits processor already has the right architecture for it.

PS: coverage check is still broken, there's an issue for that we should work on at some point

### CODE REVIEW

**Comment by RobinPicard** [comment]
The PR to add support for thinking models and forcing them to stop thinking after a given number of tokens have been used is coming soon. It will be available for all steerable models/tensor types.
